### PR TITLE
44077 First pass at support in tk-config-default2 with tk-multi-publi…

### PR DESCRIPTION
**Pull request by gomezll**
…sh2 engine hooks

related changes

https://github.com/shotgunsoftware/tk-multi-publish2/pull/38
https://github.com/shotgunsoftware/tk-config-default2/pull/15

### CODE REVIEW

File: hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py [hunk 1]
```diff
@@ -0,0 +1,431 @@
+﻿# Copyright (c) 2017 Shotgun Software Inc.
```
[the remark this answers is not in the document]
Comment (gomezll) [reply hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py]: Renamed

File: hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+
+from pyfbsdk import FBApplication
+
+mb_app = FBApplication()
+
+HookBaseClass = sgtk.get_hook_baseclass()
+
+
+class MotionBuilderSessionPublishPlugin(HookBaseClass):
+    """
```
[the remark this answers is not in the document]
Comment (gomezll) [reply hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py]: Updated

File: hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+    Plugin for publishing an open motion builder session.
+    """
+
+    @property
+    def icon(self):
```
[the remark this answers is not in the document]
Comment (gomezll) [reply hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py]: Removed

File: hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+    @property
+    def description(self):
+        """
+        Verbose, multi-line description of what the plugin does. This can
+        contain simple html for formatting.
+        """
+
+        loader_url = "https://support.shotgunsoftware.com/hc/en-us/articles/219033078"
+
+        return """
```
[the remark this answers is not in the document]
Comment (gomezll) [reply hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py]: Copied

File: hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+        The type string should be one of the data types that toolkit accepts as
+        part of its environment configuration.
+        """
+
+        # inherit the settings from the base publish plugin
+        base_settings = super(MotionBuilderSessionPublishPlugin, self).settings or {}
+
+        # settings specific to this class
+        mobu_publish_settings = {
+            "Publish Type": {
```
[the remark this answers is not in the document]
Comment (gomezll) [reply hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py]: Removed

File: hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+            self.logger.warn(
+                "The Motion Builder session has not been saved.",
+                extra=_get_save_as_action()
+            )
+
+        accepted = True
+        publisher = self.parent
+        template_name = settings["Publish Template"].value
+
+        publish_template = publisher.get_template_by_name(template_name)
```
[the remark this answers is not in the document]
Comment (gomezll) [reply hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py]: Removed

File: hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+                extra=_get_save_as_action()
+            )
+            return False
+
+        # get the path in a normalized state. no trailing separator,
+        # separators are appropriate for current os, no double separators,
+        # etc.
+        sgtk.util.ShotgunPath.normalize(path)
+
+        # get the publish name for this file path. this will ensure we get a
```
[the remark this answers is not in the document]
Comment (gomezll) [reply hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py]: removed

File: hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+                    }
+                )
+            else:
+                self.logger.debug(
+                    "Work template configured and matches session file.")
+        else:
+            self.logger.debug("No work template configured.")
+
+        # if the file has a version number in it, see if the next version exists
+        next_version_path = publisher.util.get_next_version_path(path)
```
[the remark this answers is not in the document]
Comment (gomezll) [reply hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py]: changed

File: hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py [hunk 9]
```diff
@@ -0,0 +1,431 @@
...
+        """
+        Executes the publish logic for the given item and settings.
+
+        :param settings: Dictionary of Settings. The keys are strings, matching
+            the keys returned in the settings property. The values are `Setting`
+            instances.
+        :param item: Item to process
+        """
+
+        publisher = self.parent
```
[the remark this answers is not in the document]
Comment (gomezll) [reply hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py]: removed

File: hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+
+def _get_save_as_action():
+    """
+
+    Simple helper for returning a log action dict for saving the session
+    """
+    engine = sgtk.platform.current_engine()
+
+    # default save callback
+    callback = lambda: _save_session(_session_path())
```
[the remark this answers is not in the document]
Comment (gomezll) [reply hooks/tk-multi-publish2/basic/publish_motionbuilder_session.py]: http://docs.autodesk.com/MB/MB2013/ENU/MotionBuilder-SDK-Documentation/index.html It's a Save As command

File: hooks/tk-multi-publish2/basic/start_version_control.py [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+        :param item: Item to process
+
+        :returns: dictionary with boolean keys accepted, required and enabled
+        """
+
+        publisher = self.parent
+        path = _session_path()
+
+        if path:
+            version_number = publisher.util.get_version_number(path)
```
[the remark this answers is not in the document]
Comment (gomezll) [reply hooks/tk-multi-publish2/basic/start_version_control.py]: Changed

File: hooks/tk-multi-publish2/basic/start_version_control.py [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+
+        if not path:
+            # the session still requires saving. provide a save button.
+            # validation fails
+            self.logger.error(
+                "The Motion Builder session has not been saved.",
+                extra=_get_save_as_action()
+            )
+            return False
+
```
[the remark this answers is not in the document]
Comment (gomezll) [reply hooks/tk-multi-publish2/basic/start_version_control.py]: Added comment

File: hooks/tk-multi-publish2/basic/start_version_control.py [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+        be used to version up files.
+
+        :param settings: Dictionary of Settings. The keys are strings, matching
+            the keys returned in the settings property. The values are `Setting`
+            instances.
+        :param item: Item to process
+        """
+        pass
+
+
```
[the remark this answers is not in the document]
Comment (gomezll) [reply hooks/tk-multi-publish2/basic/start_version_control.py]: added

File: hooks/tk-multi-publish2/basic/start_version_control.py [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+def _save_session(path):
+    """
+    Save the current session to the supplied path.
+    """
+
+    mb_app.FileSave(path)
+
+
+def _get_save_as_action():
```
[the remark this answers is not in the document]
Comment (gomezll) [reply hooks/tk-multi-publish2/basic/start_version_control.py]: Updated

File: hooks/tk-multi-publish2/basic/start_version_control.py [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+    """
+    Simple helper for returning a log action to show version docs
+    """
+    return {
+        "action_open_url": {
+            "label": "Version Docs",
+            "tooltip": "Show docs for version formats",
+            "url": "https://support.shotgunsoftware.com/hc/en-us/articles/115000068574-User-Guide-WIP-#What%20happens%20when%20you%20publish"
+        }
+    }
```
[the remark this answers is not in the document]
Comment (gomezll) [reply hooks/tk-multi-publish2/basic/start_version_control.py]: updated

File: hooks/tk-multi-publish2/basic/start_version_control.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (josh-t) [reply hooks/tk-multi-publish2/basic/start_version_control.py]: Is there a way to show a "Save As" dialog for Mobu programatically? The fallback you have is to call `_save_session()` with a path argument. The code that calls this method is typically in a situation where the current session path is invalid or non existent so this won't work.

File: hooks/tk-multi-publish2/basic/publish_session.py [hunk 16]
```diff
@@ -0,0 +1,392 @@
...
+
+def _get_save_as_action():
+    """
+
+    Simple helper for returning a log action dict for saving the session
+    """
+    engine = sgtk.platform.current_engine()
+
+    # default save callback
+    callback = lambda: _save_session(_session_path())
```
Comment (josh-t) [inline hooks/tk-multi-publish2/basic/publish_session.py]: Same comment here about the fallback "Save As..." requirement.

File: hooks/tk-multi-publish2/basic/publish_session.py [hunk 17]
```diff
@@ -12,7 +12,7 @@
 import pprint
 import sgtk
 
-from pyfbsdk import FBApplication
+from pyfbsdk import *
```
Comment (josh-t) [inline hooks/tk-multi-publish2/basic/publish_session.py]: Importing `*` is discouraged. I suggest importing what you need explicitly:

`from pyfbsdk import FBApplication, FBFilePopup, FBFilePopupStyle`

**Pull request merged**